Add Job Sample Automation

labkey-danield opened the pull request:
I'm still working on test cases but the support libs should be good to review.
Several of the changes in the repo are there just to get the story done, some of the modules I changed need to go away and be replaced with a shared component between Biologics and SampleManager and the other parts of the world.

labkey-danield committed:
Merge branch 'develop' into fb_addJobSampleAutomation

labkey-danield committed:
Merge branch 'develop' into fb_addJobSampleAutomation

labkey-danield committed:
Adding automation for add samples to job.
Fixing issues with the finders for the QueryGrid.
Updated other tests that were impacted by the finder changes.
Some other clean-up and added comments.

labkey-danield committed:
Merge branch 'develop' into fb_addJobSampleAutomation

labkey-tchad commented on src/org/labkey/test/components/glassLibrary/grids/GridBar.java:
```diff
@@ -30,9 +30,9 @@
+    protected GridBar(WebDriver driver, WebElement container, ResponsiveGrid responsiveGrid)
     {
-        _gridBarElement = buttonBar;
+        _gridBarElement = Locator.tagWithClass("div", "query-grid-bar").findElement(container);
```
This Locator is defined below. Also, use `findWhenNeeded` so that this component can be initialized lazily:
```suggestion
        _gridBarElement = Locators.gridBar().findWhenNeeded(container);
```

labkey-danield replied:
Updated.

labkey-tchad commented on src/org/labkey/test/components/glassLibrary/grids/QueryGrid.java:
```diff
@@ -52,18 +50,8 @@ public WebDriver getDriver()
-    public boolean hasGrid()
-    {
-        return _responsiveGrid.isPresent();
+        _responsiveGrid = new ResponsiveGrid.ResponsiveGridFinder(_driver).find(_queryGridPanel);
+        return _responsiveGrid;
```
`_responsiveGrid` was set in the constructor. We should be able to just return it here.
```suggestion
        return _responsiveGrid;
```

labkey-danield replied:
The call to the finder was there originally (so to speak). The only thing that really changed here was to no longer using optionals. 
I remember we discussed this way back and landed on calling the finder when the grid is requested because it covers the scenario of the grid being updated between the constructor and when this function might be called.
An alternative could be to check for staleness and only call the finder then, otherwise return the grid found int he constructor.

labkey-tchad reviewed (changes requested):
Where did we come down on renaming `QueryGrid` to `QueryGridPanel`? I think having panel in the name makes it a bit more clear what the component represents. Maybe our decision was to do that refactor on its own in a separate branch to make the code review for this a bit more straightforward?

I think we could simplify and clarify the hierarchy of these components. I would suggest having `GridBar` and `ResponsiveGrid` take only a `QueryGrid` in their constructor. That would be the `container` element, provide a `WebDriver` instance (need to make `QueryGrid.getDriver` public),  and provide the `GridBar` and `ResponsiveGrid` access to each-other via the getters in `QueryGrid`.
If we have a scenario in the future where there is a `GridBar` in a different context, we can pull the functionality into a base class that doesn't require a `QueryGrid`.

labkey-danield committed:
Merging develop into fb_addJobSampleAutomation.

labkey-danield committed:
Adding more test cases.
Implementing suggestions from code review feedback.
Moved the attached samples to their own component (panel) to avoid confusion with search samples component.
Renamed some functions for clarity.

labkey-danield committed:
Reverting change that regressed table functionality.
